pennylane/pauli/pauli_arithmetic.py

```diff
@@ -964,6 +946,24 @@ def _get_same_structure_csr(self, pauli_words, wire_order):
         data = outer.T @ inner
         return indices, data.ravel()
 
+    def _sum_same_structure_pws_dense(self, pauli_words, wire_order):
+        matrix_size = 2 ** (len(wire_order))
+        base_matrix = sparse.csr_matrix((matrix_size, matrix_size), dtype="complex128")
+
+        data0 = pauli_words[0]._get_csr_data(wire_order, 1)
+        base_matrix.data = np.ones_like(data0)
+        base_matrix.indices = pauli_words[0]._get_csr_indices(wire_order)
+        base_matrix.indptr = _cached_arange(
+            matrix_size + 1
+        )  # Non-zero entries by row (starting from 0)
+        base_matrix = base_matrix.toarray()
+        coeff = self[pauli_words[0]]
+        data = coeff * qml.math.convert_like(data0, coeff)
+        for pw in pauli_words[1:]:
+            data += self[pw] * pw._get_csr_data(wire_order, 1)
+
+        return qml.math.einsum("ij,i->ij", base_matrix, data)
```

Comment:
Smart, I like it.

Reply:
Well, large matrices and vectors is what we care about here.

Reply:
Tried substituting `o2` and got a bunch of test failures.

Reply:
would `qml.math.einsum("i, ij ->ij", data, base_matrix)` be any faster?

Reply:
> would `qml.math.einsum("i, ij ->ij", data, base_matrix)` be any faster?

That appears like the correct `o2` one.

Reply:
I would think `einsum` is smart enough to handle both in an optimal way, such that they come out similar speed. `o1` should be a loop over scaling rows and `o2` should be a loop over vector-multiply. `o1` has fewer memory accesses in the limit of long rows. `o2` might be better for short rows.

Reply:
I'm just going to leave it be for now, but we could probably come back to this line at some point.